Fix verified user getting email verifications

yatharth1706 opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to appwrite here: https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## What does this PR do?

Earlier when email verification post request is sent for already verified user, email is sent again. Its similar in phone verification as well. I have written the logic to check first whether email is verified or not. If yes, throw the error that email is already verified. Similarly did for phone verification post API.

Fixes: https://github.com/appwrite/appwrite/issues/4797 

## Test Plan

None. I would appreciate if you guys could help me test this. I am not sure how i will test this locally. What steps need to be done?

## Related PRs and Issues

* https://github.com/appwrite/appwrite/issues/4797

## Checklist

- [X] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

yatharth1706 commented:
@stnguyen90 , @eldadfux can anyone review this. If the logic is fine?

zvikfir commented on app/controllers/api/account.php:
```diff
@@ -2200,6 +2200,10 @@
             throw new Exception(Exception::GENERAL_SMTP_DISABLED, 'SMTP Disabled');
         }
+        
+        if($user->getAttribute('emailVerification') == true){
```
The comparison of a boolean to a `true` value is redundant.

yatharth1706 replied:
fixed

zvikfir commented on app/controllers/api/account.php:
```diff
@@ -2200,6 +2200,10 @@
             throw new Exception(Exception::GENERAL_SMTP_DISABLED, 'SMTP Disabled');
         }
+        
+        if($user->getAttribute('emailVerification') == true){
```
Use a linter. You're missing a whitespace after the `if` keyword.
```suggestion
        if ($user->getAttribute('emailVerification') == true) {
```

yatharth1706 replied:
done

zvikfir commented on src/Appwrite/Extend/Exception.php:
```diff
@@ -73,6 +73,8 @@ class Exception extends \Exception
     public const USER_PHONE_ALREADY_EXISTS         = 'user_phone_already_exists';
     public const USER_PHONE_NOT_FOUND              = 'user_phone_not_found';
     public const USER_MISSING_ID                   = 'user_missing_id';
+    public const USER_EMAIL_ALREADY_VERIFIED        = 'user_email_already_verified';
+    public const USER_PHONE_ALREADY_VERIFIED        = 'user_phone_already_verified';
```
Values are not aligned with the rest.

```suggestion
    public const USER_EMAIL_ALREADY_VERIFIED       = 'user_email_already_verified';
    public const USER_PHONE_ALREADY_VERIFIED       = 'user_phone_already_verified';
```

yatharth1706 replied:
fixed

yatharth1706 commented on app/config/errors.php:
```diff
@@ -190,6 +190,16 @@
+    Exception::USER_EMAIL_ALREADY_VERIFIED => [
+        'name' => Exception::USER_EMAIL_ALREADY_VERIFIED,
+        'description' => 'User email is already verified',
+        'code' => 400,
```
done

stnguyen90 reviewed (changes requested):
@yatharth1706 would you please add a test cases for this?

yatharth1706 commented:
> @yatharth1706 would you please add a test cases for this?

Sure @stnguyen90, I will try to add test case for this. Will let u know once done

yatharth1706 commented:
> > @yatharth1706 would you please add a test cases for this?
> 
> Sure @stnguyen90, I will try to add test case for this. Will let u know once done

Hy @stnguyen90 , can you guide me on the process to create new tests. I am not sure how to add tests in this and verify if everythings working fine

stnguyen90 commented:
> can you guide me on the process to create new tests. I am not sure how to add tests in this and verify if everythings working fine

Look through https://github.com/appwrite/appwrite/blob/master/tests/e2e/Services/Account/AccountBase.php to find the relevant place to add the test. To run tests, refer to https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md#tests

yatharth1706 commented:
> > can you guide me on the process to create new tests. I am not sure how to add tests in this and verify if everythings working fine
> 
> Look through https://github.com/appwrite/appwrite/blob/master/tests/e2e/Services/Account/AccountBase.php to find the relevant place to add the test. To run tests, refer to https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md#tests

@stnguyen90, I have added test cases for email verification and phone verification, both are tested locally. They are working fine. Attaching screenshot for reference. Please review the code. Let me know if anything is missing. 

![Screenshot from 2023-05-07 08-46-08](https://user-images.githubusercontent.com/32243289/236655905-40055dee-d495-4cd9-89a5-a746cada29c4.png)

stnguyen90 commented on tests/e2e/Services/Account/AccountBase.php:
```diff
@@ -931,6 +932,58 @@ public function testUpdateAccountVerification($data): array
+        /**
+         * Test for FAILURE
+         */
+        $response = $this->client->call(Client::METHOD_POST, '/account/verification', array_merge([
+            'origin' => 'http://localhost',
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'cookie' => 'a_session_' . $this->getProject()['$id'] . '=' . $session,
+        ]), [
+            'url' => 'localhost/verification',
+        ]);
+
+        $this->assertEquals(400, $response['headers']['status-code']);
```
This is an error because the url is invalid? We have this test elsewhere so no need to do that here.

stnguyen90 commented on tests/e2e/Services/Account/AccountBase.php:
```diff
@@ -931,6 +932,58 @@ public function testUpdateAccountVerification($data): array
+
+        $this->assertEquals(400, $response['headers']['status-code']);
+
+        $response = $this->client->call(Client::METHOD_POST, '/account/verification', array_merge([
+            'origin' => 'http://localhost',
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'cookie' => 'a_session_' . $this->getProject()['$id'] . '=' . $session,
+        ]), [
+            'url' => 'http://remotehost/verification',
+        ]);
+
+        $this->assertEquals(400, $response['headers']['status-code']);
```
This is also the same as another test elsewhere, right?

stnguyen90 replied:
> i want error code to be 400 if email or phone no verification is called again after successful verification. Thats why two times function tests are there

But it shouldn't be 400. It should be 409 and you already have that test above 🧐 

> I saw that verification test was here so i added it

Ya, here is fine since there are already /account/verification API calls in here. And the phone verification one in the other one is good too.

stnguyen90 commented on tests/e2e/Services/Account/AccountCustomClientTest.php:
```diff
@@ -1044,4 +1044,28 @@ public function testUpdatePhoneVerification($data): array
+
+        $this->assertEquals(409, $response['headers']['status-code']);
+
+        return \array_merge($data);
```
no need to call `\array_merge()` since you're not merging anything.

stnguyen90 commented:
@yatharth1706, will you be able to address the comments from the review?

yatharth1706 commented:
> @yatharth1706, will you be able to address the comments from the review?

Sure @stnguyen90 , I will try to do these by today or tomorrow. Earlier I didn't get time because of appwrite hackathon. 
I will try to address those comments by tomorrow max.

stnguyen90 commented:
@yatharth1706, did all the tests pass for you when you ran it locally?

yatharth1706 commented:
> @yatharth1706, did all the tests pass for you when you ran it locally?

Yes @stnguyen90 I ran the following command on gitpod when i was doing development 

docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account

All tests ran fine locally.

stnguyen90 commented:
> docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account

@yatharth1706, what about if you run all the tests?

```
docker compose exec appwrite test
```

stnguyen90 reviewed (commented):
> docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account

@yatharth1706, what about if you run all the tests?

```
docker compose exec appwrite test
```

yatharth1706 commented:
> > docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account
> 
> @yatharth1706, what about if you run all the tests?
> 
> ```
> docker compose exec appwrite test
> ```

while running all tests, one test is failing. I'll check this

yatharth1706 commented:
> > docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account
> 
> @yatharth1706, what about if you run all the tests?
> 
> ```
> docker compose exec appwrite test
> ```

The reason it might be failing because i have already written code in such a way that two times account verification for same user wont work. Second time it will give error. That has already been tested in AccountBase.php tests file. But in webhooks again verification test is there. So its failing. Need to think about it i guess

stnguyen90 commented:
> The reason it might be failing because i have already written code in such a way that two times account verification for same user wont work. Second time it will give error. That has already been tested in AccountBase.php tests file. But in webhooks again verification test is there. So its failing. Need to think about it i guess

Maybe you can update the webhook one to trigger the verification on a different account? Or update the account test to trigger on a different account?

I'll set this to draft while you work on the change.

yatharth1706 commented:
> > The reason it might be failing because i have already written code in such a way that two times account verification for same user wont work. Second time it will give error. That has already been tested in AccountBase.php tests file. But in webhooks again verification test is there. So its failing. Need to think about it i guess
> 
> Maybe you can update the webhook one to trigger the verification on a different account? Or update the account test to trigger on a different account?
> 
> I'll set this to draft while you work on the change.

hi @stnguyen90, I fixed the issue. Can you once review it. I found out one problem in the webhooks test before verification account recovery apis were getting called and i found out that it was changing the verification status of the user. Thats why account/verification test was failing as the email was already verified. I shifted those functions after the verification api calls and now its working fine. I tested locally. 

Can you review it once if thats fine

stnguyen90 commented:
@yatharth1706 looks like there are merge conflicts now. Would you please resolve those conflicts and then mark the PR as Ready for review so it's no longer a draft?

yatharth1706 commented:
> @yatharth1706 looks like there are merge conflicts now. Would you please resolve those conflicts and then mark the PR as Ready for review so it's no longer a draft?

Hi @stnguyen90 , can you review this. I have fixed all the merge conflicts. I also tested the test cases locally. They are working fine. But here, tests are failing in e2e/Services/Account

But i ran this command locally
docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Account

Every test case worked fine. Not sure what's the issue. Can you look at this once ?

<img width="921" alt="Screenshot 2023-09-23 at 12 20 14 AM" src="https://github.com/appwrite/appwrite/assets/32243289/788287c6-b552-4d02-9aed-7e3d8d73bdd9">

yatharth1706 commented:
Hi @stnguyen90, this time i ran full test suite using the below command 
```docker compose exec -T appwrite test --debug   ```

And all tests ran fine, except one test in location Tests\E2E\Services\Databases\DatabasesConsoleClientTest::testGetCollectionLogs

I dont know the reason and also tests are so inconsistent. Like in git when tests are running some other test case is getting failed and locally some other is failing

Although my tests related to account one ran successfully locally. Can you please take a look and review the code once.

Attaching screenshot of tests suite which i tested locally as well for reference:
![image](https://github.com/appwrite/appwrite/assets/32243289/44e6b4ee-3b52-4017-9a44-fbbd97943a6e)

stnguyen90 commented on app/assets/security 2/10k-common-passwords:
```diff
@@ -0,0 +1,10000 @@
+password
```
This file shouldn't be in this PR, should it?

yatharth1706 replied:
I guess because of merge, this might have come. Will look into this.

stnguyen90 commented on app/controllers/shared/api 2/auth.php:
```diff
@@ -0,0 +1,62 @@
+<?php
```
This file shouldn't be in this PR, should it?

stnguyen90 commented on src/Appwrite/Extend/Exception.php:
```diff
@@ -84,6 +84,8 @@ class Exception extends \Exception
     public const USER_OAUTH2_BAD_REQUEST           = 'user_oauth2_bad_request';
     public const USER_OAUTH2_UNAUTHORIZED          = 'user_oauth2_unauthorized';
     public const USER_OAUTH2_PROVIDER_ERROR        = 'user_oauth2_provider_error';
+    public const USER_EMAIL_ALREADY_VERIFIED        = 'user_email_alread_verified';
+    public const USER_PHONE_ALREADY_VERIFIED        = 'user_phone_already_verified';
```
Let's make sure the alignment is consistent.

```suggestion
    public const USER_EMAIL_ALREADY_VERIFIED       = 'user_email_alread_verified';
    public const USER_PHONE_ALREADY_VERIFIED       = 'user_phone_already_verified';
```

yatharth1706 replied:
Sure will fix this

yatharth1706 commented:
@stnguyen90 , can you review this?

stnguyen90 reviewed (changes requested):
@yatharth1706. tests are failing because your branch might be behind. Would you be able to update?

yatharth1706 commented:
Hi @stnguyen90, I have updated with the main branch, but still some test cases are failing for e2e/Services/Avatars. I dont know the reason. I tried running whole tests and seperate tests as well for Avatars

docker compose exec appwrite test /usr/src/code/tests/e2e/Services/Avatars

6 tests are failing. Do you know the reason, why it could be failing. I have updated my branch using git merge master after sync my forked master with appwrite main branch.

stnguyen90 commented on tests/e2e/Services/Account/AccountBase.php:
```diff
@@ -891,6 +891,7 @@ public function testCreateAccountVerification($data): array
         return $data;
     }
 
+
```
Please remove the empty line

yatharth1706 replied:
Done

stnguyen90 commented on tests/e2e/Services/Account/AccountBase.php:
```diff
@@ -945,6 +946,33 @@ public function testUpdateAccountVerification($data): array
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'cookie' => 'a_session_' . $this->getProject()['$id'] . '=' . $session,
+
```
Please remove the empty line

yatharth1706 replied:
Done

stnguyen90 commented on tests/e2e/Services/Account/AccountCustomClientTest.php:
```diff
@@ -1068,4 +1068,28 @@ public function testUpdatePhoneVerification($data): array
+            'content-type' => 'application/json',
+            'x-appwrite-project' => $this->getProject()['$id'],
+            'cookie' => 'a_session_' . $this->getProject()['$id'] . '=' . $session,
+
```
Please remove the empty line

yatharth1706 replied:
Done

stnguyen90 reviewed (changes requested):
Some minor fixes.

Also, the avatars fix was added in 1.4.x branch. Do you think you'd be able to remove commits from your branch that were in main and then rebase off of 1.4.x. The goal is that when we merge into 1.4.x, the PR should only contain your changes and the tests should pass.

Sorry for the inconvenience 🙈

yatharth1706 commented:
> Some minor fixes.
> 
> Also, the avatars fix was added in 1.4.x branch. Do you think you'd be able to remove commits from your branch that were in main and then rebase off of 1.4.x. The goal is that when we merge into 1.4.x, the PR should only contain your changes and the tests should pass.
> 
> Sorry for the inconvenience 🙈

Hi @stnguyen90 , i completed changes suggested by you. Can you review

stnguyen90 reviewed (commented):
Does app/console need to be in this PR? If not, it would be good to remove it to reduce chances of merge conflicts.

yatharth1706 commented:
@stnguyen90 , I have removed the console directory. Please review once.

stnguyen90 commented:
@yatharth1706, it seems like the console repo was deleted entirely rather than just removed from the PR 🧐

stnguyen90 reviewed (changes requested):
@yatharth1706, it seems like the console repo was deleted entirely rather than just removed from the PR 🧐 The change has lead to a merge conflict now.

yatharth1706 commented:
> Reference



> @yatharth1706, it seems like the console repo was deleted entirely rather than just removed from the PR 🧐 The change has lead to a merge conflict now.

Hi @stnguyen90 , I am little confused. You told to remove it. So i deleted it i guess. Also those changes in app/console must have come when i merged it with the base branch i guess.

Can you tell about how should i go about this regarding removing this from PR ? Like how does that actually happens.

stnguyen90 commented:
@yatharth1706 you can try going into the folder and checking out the tag that it should be. then, commit that.

yatharth1706 commented:
> @yatharth1706 you can try going into the folder and checking out the tag that it should be. then, commit that.

Hi @stnguyen90  can you review now, I have checked out the latest tag in console submodule. Please review

stnguyen90 commented:
@yatharth1706, sorry, app/console is still in the PR: 

<img width="1211" alt="image" src="https://github.com/appwrite/appwrite/assets/1477010/21fa4fd1-bc1a-4315-840e-d2f726cd7602">

Perhaps the .gitmodules file also needs to be restored.

stnguyen90 reviewed (changes requested):
@yatharth1706, sorry, app/console is still in the PR: 

<img width="1211" alt="image" src="https://github.com/appwrite/appwrite/assets/1477010/21fa4fd1-bc1a-4315-840e-d2f726cd7602">

Perhaps the .gitmodules file also needs to be restored.

yatharth1706 commented:
> @yatharth1706, sorry, app/console is still in the PR:
> 
> <img alt="image" width="1211" src="https://user-images.githubusercontent.com/1477010/280854700-21fa4fd1-bc1a-4315-840e-d2f726cd7602.png">
> Perhaps the .gitmodules file also needs to be restored.

I did exactly same steven which you told i went inside console and checked out the tag the latest tag of console repo. Its always going when i am commiting. Can you try merging? It should merge.
